### PR TITLE
feat(provider/ecs): ECS Cache base classes

### PR DESCRIPTION
**Pull request by dkirillov**
feat(provider/ecs): ECS Cache base classes

Added base classes that will be used for the `CachingAgent`s classes that are to come.

@cfieber @ajordens @robzienert @BrunoCarrier

### CODE REVIEW

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/EcsProvider.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  public void synchronizeHealthAgents() {
+    healthAgents = Collections.unmodifiableCollection(agents.stream()
+      .filter(a -> a instanceof HealthProvidingCachingAgent)
+      .map(a -> (HealthProvidingCachingAgent) a).collect(Collectors.toList()));
+  }
+
+  public Collection<HealthProvidingCachingAgent> getHealthAgents() {
+    List<HealthProvidingCachingAgent> newHealthAgents = new LinkedList<>();
+    newHealthAgents.addAll(healthAgents);
+    return Collections.unmodifiableCollection(newHealthAgents);
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/EcsProvider.java]: Couldn't you just do `return Collections.unmodifiableCollection(healthAgents);` here?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+   * @param ecs The AmazonECS client to use for querying.
+   * @param providerCache The ProviderCache to retrieve clusters from.
+   * @return A set of ECS cluster ARNs.
+   */
+  Set<String> getClusters(AmazonECS ecs, ProviderCache providerCache) {
+    Set<String> clusters = providerCache.getAll(ECS_CLUSTERS.toString()).stream()
+      .map(cacheData -> (String) cacheData.getAttributes().get("clusterArn"))
+      .collect(Collectors.toSet());
+
+    if (clusters == null || clusters.isEmpty()) {
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: `clusters` will never be `null`, will it?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+   * Currently only supports the caching agent being authoritative over one key namespace.
+   * @return Key namespace.
+   */
+  String getAuthoritativeKeyName() {
+    Collection<AgentDataType> authoritativeNamespaces = getProvidedDataTypes().stream()
+      .filter(agentDataType -> agentDataType.getAuthority().equals(AUTHORITATIVE))
+      .collect(Collectors.toSet());
+
+    if (authoritativeNamespaces.size() != 1) {
+      throw new RuntimeException("AbstractEcsCachingAgent supports only one authoritative key namespace.");
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: Perhaps including the actual size of the collection in the exception would be valuable? `AbstractEcsCachingAgent supports only one authoritative key namespace (%d given).` sorta thing

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    return authoritativeNamespaces.iterator().next().getTypeName();
+  }
+
+  CacheResult buildCacheResult(String authoritativeKeyName, List<T> items, ProviderCache providerCache) {
+    String prettyKeyName = CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, authoritativeKeyName);
+
+    Map<String, Collection<CacheData>> dataMap = generateFreshData(items);
+
+    Set<String> oldKeys = providerCache.getAll(authoritativeKeyName).stream()
+      .map(cache -> cache.getId()).collect(Collectors.toSet());
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: Drop `.collect()` to a new line.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+  /**
+   * Evicts cache that does not belong to an entity on the ECS service.
+   * This is done by evicting old keys that are no longer found in the new keys provided by the new data.
+   * @param newData New data that contains new keys.
+   * @param oldKeys Old keys.
+   * @return Key collection associated to the key namespace the the caching agent is authoritative of.
+   */
+  private Map<String, Collection<String>> computeEvictableData(Collection<CacheData> newData, Collection<String> oldKeys) {
+    Set<String> newKeys = newData.stream().map(newKey -> newKey.getId()).collect(Collectors.toSet());
```
Comment (robzienert) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: `newData.stream().map(CacheData::getId).collect(Collectors.toSet())` is a bit cleaner.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/EcsProvider.java (hunk 1 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/EcsProvider.java]: Changed.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java (hunk 2 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: Changed - removed `clusters == null`.

You're right - it should never be null. 
`providerCache.getAll(..)` will return an empty set, and the stream will also return an empty set if nothing filters through.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java (hunk 3 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: Done.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java (hunk 4 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: Done.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java (hunk 5 above)
Comment (dkirillov) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java]: Done.

**Comment by dkirillov** [comment]
I've made the changes @robzienert

**Pull request merged**